cpp/src/arrow/compute/kernels/vector_dictionary_test.cc

```diff
@@ -0,0 +1,139 @@
+                               const std::string& input_dictionary_json,
+                               const std::string& input_index_json,
+                               const std::string& expected_dictionary_json,
+                               const std::string& expected_index_json) {
+  auto input = DictArrayFromJSON(dict_type, input_index_json, input_dictionary_json);
+  auto expected =
+      DictArrayFromJSON(dict_type, expected_index_json, expected_dictionary_json);
+
+  ASSERT_OK_AND_ASSIGN(Datum actual_datum, DictionaryCompaction(input));
+  ValidateOutput(actual_datum);
+  std::shared_ptr<Array> actual = actual_datum.make_array();
+  AssertArraysEqual(*expected, *actual, /*verbose=*/true);
+}
+
+void CheckDictionaryCompactionOnChunks(const std::shared_ptr<DataType>& dict_type,
+                                       const ArrayVector& input,
+                                       const ArrayVector& expected) {
+  auto input_chunked_array = std::make_shared<ChunkedArray>(input, dict_type);
+
+  ASSERT_OK_AND_ASSIGN(Datum actual_datum, DictionaryCompaction(input_chunked_array));
+  ValidateOutput(actual_datum);
+  auto actual = actual_datum.chunked_array();
+  AssertChunkedEqual(*actual, expected);
+}
+
+TEST_F(TestDictionaryCompactionKernel, DictionaryArray) {
+  std::shared_ptr<arrow::DataType> type;
+  std::shared_ptr<arrow::DataType> dict_type;
+
+  for (const auto& index_type : all_dictionary_index_types()) {
+    ARROW_SCOPED_TRACE("index_type = ", index_type->ToString());
+
+    type = boolean();
+    dict_type = dictionary(index_type, type);
+    // input is compacted
+    CheckDictionaryCompaction(dict_type, "[]", "[]", "[]", "[]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0, 1, 0]", "[true, false]",
+                              "[0, 1, 0]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, 1, 0]",
+                              "[true, null, false]", "[2, 1, 0]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0, null, 1, 0]",
+                              "[true, false]", "[0, null, 1, 0]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, null, 1, 0]",
+                              "[true, null, false]", "[2, null, 1, 0]");
+    // input isn't compacted
+    CheckDictionaryCompaction(dict_type, "[null]", "[]", "[]", "[]");
+    CheckDictionaryCompaction(dict_type, "[false]", "[null]", "[]", "[null]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0]", "[true]", "[0]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, 1]", "[null, false]",
+                              "[1, 0]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0, null]", "[true]",
+                              "[0, null]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, null, 1]",
+                              "[null, false]", "[1, null, 0]");
+  }
```

Comment:
It'd be nice to also add a test case with a larger input array, where several contiguous chunks of the dictionary array are unused. It'd stress the logic for index re-adjustment more.

Reply:
Sure, I add some tests and a comment for that.